src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,594 @@
+                              num_samples=self.num_samples,
+                              random_fraction=self.random_fraction,
+                              bandwidth_factor=self.bandwidth_factor,
+                              min_bandwidth=self.min_bandwidth)
+        else:
+            raise ValueError('Error: Search space is None')
+        # generate first brackets
+        self.generate_new_bracket()
+        send(CommandType.Initialized, '')
+        return True
+
+    def generate_new_bracket(self):
+        """generate a new bracket"""
+        logger.debug(
+            'start to create a new SuccessiveHalving iteration, self.curr_s=%d', self.curr_s)
+        if self.curr_s < 0:
+            logger.info("s < 0, Finish this round of Hyperband in BOHB. Generate new round")
+            self.curr_s = self.s_max
+        self.brackets[self.curr_s] = Bracket(s=self.curr_s, s_max=self.s_max, eta=self.eta,
+                                    max_budget=self.max_budget, optimize_mode=self.optimize_mode)
+        next_n, next_r = self.brackets[self.curr_s].get_n_r()
+        logger.debug(
+            'new SuccessiveHalving iteration, next_n=%d, next_r=%d', next_n, next_r)
+        # rewrite with TPE
+        generated_hyper_configs = self.brackets[self.curr_s].get_hyperparameter_configurations(
+            next_n, next_r, self.cg)
+        self.generated_hyper_configs = generated_hyper_configs.copy()
+
+    def handle_request_trial_jobs(self, data):
+        """recerive the number of request and generate trials
+
+        Parameters
+        ----------
+        data: int
+            number of trial jobs that nni manager ask to generate
+        """
+        # Receive new request
+        self.credit += data
+
+        for _ in range(self.credit):
+            self._request_one_trial_job()
+
+        return True
+
+    def _request_one_trial_job(self):
+        """get one trial job, i.e., one hyperparameter configuration.
+
+        Returns
+        -------
+        dict:
+            one hyperparameter configuration
+            0: 'parameter_id', id of new hyperparameter
+            1: 'parameter_source', 'algorithm'
+            2: 'parameters', value of new hyperparameter
+        """
+        if not self.generated_hyper_configs:
+            ret = {
+                'parameter_id': '-1_0_0',
+                'parameter_source': 'algorithm',
+                'parameters': ''
+            }
+            send(CommandType.NoMoreTrialJobs, json_tricks.dumps(ret))
+            return True
+        assert self.generated_hyper_configs
+        params = self.generated_hyper_configs.pop()
+        ret = {
+            'parameter_id': params[0],
+            'parameter_source': 'algorithm',
+            'parameters': params[1]
+        }
+        self.parameters[params[0]] = params[1]
+        send(CommandType.NewTrialJob, json_tricks.dumps(ret))
+        self.credit -= 1
+        return True
```

Comment:
same here. Seems only return True in this function?   But in the docstring....

       Returns
        -------
        dict:
            one hyperparameter configuration
            0: 'parameter_id', id of new hyperparameter
            1: 'parameter_source', 'algorithm'
            2: 'parameters', value of new hyperparameter
        """

Reply:
Here we use "send(CommandType.NewTrialJob, json_tricks.dumps(ret))" to return the parameters, it's a dict.

Reply:
But the docstring is to explain about this function, not "send(CommandType.NewTrialJob, json_tricks.dumps(ret))" ?

Reply:
Modified. Now put the "return command" to definition of this function.